File: packages/recognizer/android/src/main/java/com/clarionhq/recognizer/RecognizerTypes.kt
package com.clarionhq.recognizer

/**
 * Internal error type thrown by the Android RecognizerSession.
 * Mapped to NativeRecognizerError at the Nitro bridge layer.
 */
internal class RecognizerError(
  val code: String,
  message: String,
  val recoverable: Boolean = false,
  cause: Throwable? = null,
) : Exception(message, cause)

/** Word-level segment. Android natively doesn't expose these — kept for parity. */
internal data class RecognizerSegment(
  val text: String,
  val startMs: Double,
  val durationMs: Double,
  /** -1 when unknown. */
  val confidence: Double,
  val alternatives: List<String>,
)

internal data class RecognizerTranscript(
  val id: String,
  val sessionId: String,
  val timestamp: Double,
  val text: String,
  val isFinal: Boolean,
  val language: String,
  /** -1 when unknown. */
  val confidence: Double,
  /** -1 when unknown. */
  val offsetMs: Double,
  /** -1 — Android's SpeechRecognizer doesn't report this. */
  val durationMs: Double,
  /** Empty — Android doesn't report per-word segments. */
  val segments: List<RecognizerSegment>,
)

internal interface RecognizerCallbacks {
  fun onState(state: String)
  fun onAudioLevel(rms: Double, peak: Double)
  fun onPartial(transcript: RecognizerTranscript)
  fun onFinal(transcript: RecognizerTranscript)
  fun onError(error: RecognizerError)
}

File: packages/recognizer/ios/RecognizerTypes.swift
import Foundation

/// Internal error type thrown by the iOS RecognizerSession.
/// Mapped to NativeRecognizerError at the Nitro bridge layer.
internal struct RecognizerError: Error, CustomStringConvertible {
  let code: String
  let message: String
  let recoverable: Bool

  init(code: String, message: String, recoverable: Bool = false) {
    self.code = code
    self.message = message
    self.recoverable = recoverable
  }

  var description: String { "[\(code)] \(message)" }
}

internal struct RecognizerConfig {
  let language: String
  let emitPartials: Bool
  let emitAudioLevel: Bool
  let audioLevelIntervalMs: Double
  let preferOnDevice: Bool
}

internal struct RecognizerSegment {
  let text: String
  let startMs: Double
  let durationMs: Double
  let confidence: Double  // -1 when unknown
  let alternatives: [String]
}

internal struct RecognizerTranscript {
  let id: String
  let sessionId: String
  let timestamp: Double
  let text: String
  let isFinal: Bool
  let language: String
  let confidence: Double      // -1 when unknown
  let offsetMs: Double        // -1 when unknown
  let durationMs: Double      // -1 when unknown
  let segments: [RecognizerSegment]
}

internal protocol RecognizerCallbacks: AnyObject {
  func onState(_ state: String)
  func onAudioLevel(rms: Double, peak: Double)
  func onPartial(_ transcript: RecognizerTranscript)
  func onFinal(_ transcript: RecognizerTranscript)
  func onError(_ error: RecognizerError)
}
